### PR TITLE
[ci] put more logs in a folded group

### PR DESCRIPTION
**Pull request by clee2000**
fixes: request to not print the entire log file, but the last couple of lines since they are probably the most relevant

all but last 300 lines of failing tests get put into a folded group
example https://github.com/pytorch/pytorch/actions/runs/3177200444/jobs/5177703202

### CODE REVIEW

File: test/run_test.py [hunk 1]
```diff
@@ -677,11 +677,18 @@ def run_doctests(test_module, test_directory, options):
...
 def print_log_file(test: str, file_path: str, failed: bool) -> None:
+    num_lines = sum(1 for _ in open(file_path, 'rb'))
+    n = 300
     with open(file_path, "r") as f:
         print_to_stderr("")
         if failed:
-            print_to_stderr(f"PRINTING LOG FILE of {test} ({file_path})")
-            print_to_stderr(f.read())
+            print_to_stderr(f"Expand the folded group to see the beginning of the log file of {test}")
+            print_to_stderr(f"##[group]PRINTING BEGINNING OF LOG FILE of {test} ({file_path})")
```
Comment (huydhn) [inline test/run_test.py:686]: Nit: The group print is redundant when n < num_lines here, but it should be a very rare case I guess, i.e. failing right at the beginning

**Review by huydhn — APPROVED** [review]

**Review by ZainRizvi — CHANGES_REQUESTED** [review]
I think the request was a bit different. The ask was to 1) not fold lines that are currently folded, specifically 2) not fold the last 100 lines above the error message line (not from the bottom of the log file)

If you look at the below screenshot from github, the `functorch-test_ops` test failed in the middle of the log file, but only shows the last 10 lines of the log.  Quansight was asking for a feature where they don't have to expand that text.

(My above comment is based on assumption that those error lines don't always show up on the bottom, please let me know if that's wrong)

image

File: test/run_test.py [hunk 2]
```diff
@@ -677,11 +677,18 @@ def run_doctests(test_module, test_directory, options):
 
 
 def print_log_file(test: str, file_path: str, failed: bool) -> None:
+    num_lines = sum(1 for _ in open(file_path, 'rb'))
```
Comment (lezcano) [inline test/run_test.py:680]: Incredible that there's no better way to do this in Python.

**Review by lezcano — COMMENTED** [review]
Do you have an example output with this new formatting? That way it would be easier to assess how this change looks like.

Also, the error may happen in any of the different folded groups. What does it happen in that case?

**Comment by clee2000** [comment]
> Do you have an example output with this new formatting? That way it would be easier to assess how this change looks like.

I linked a run in the PR body that should show what happens in failure cases, any of the successful runs on this PR should show what a success looks like 

> Also, the error may happen in any of the different folded groups. What does it happen in that case?

CI should short circuit when a test fails so there should only be one failing/unfolded group.  If it didn't stop at the first failure, then each of the failing tests should unfold only the last 300 lines

**Review by lezcano — APPROVED** [review]
Cool! I wonder whether 300 will be too much in practice, and whether 100 lines should be enough. Do you have any example in which having 300 LOC is useful?

In any case, I love this one. Thank you!

**Comment by clee2000** [comment]
> Cool! I wonder whether 300 will be too much in practice, and whether 100 lines should be enough. Do you have any example in which having 300 LOC is useful?

300 was just a number I randomly chose, I remember seeing some really long error logs because they printed the entire tensor or something, so I erred on the side of caution

**Comment by lezcano** [comment]
IMO 100 should be more than enough even for large tensor outputs, like when `test_ops_gradients.py` fails. This should cover most cases I can think of.
The user can always expand the log, which expands like 50 lines at a time, or see the raw logs if they need to.

**Review by ZainRizvi — APPROVED** [review]

**Pull request closed without merging**